# Carregando os pacotes

In [1]:
# Importando as bibliotecas necessárias
import pandas as pd
import numpy as np
import re
#import matplotlib.pyplot as plt
#import seaborn as sns

In [2]:
# Configurações gerais
pd.set_option("display.max_columns", None) #Mostrar todas as colunas
#sns.set(style="whitegrid") # Estilo para gráficos

# Carga e compreensão dos dados

In [3]:
# Carregando os datasets
df_netflix = pd.read_csv("dataset/netflix_titles.csv")
df_disney = pd.read_csv("dataset/disney_plus_titles.csv")
df_amazon = pd.read_csv("dataset/amazon_prime_titles.csv")
df_hulu = pd.read_csv("dataset/hulu_titles.csv")

In [4]:
# Exibir uma amostra dos dados
print("Amostra - Netflix")
display(df_netflix.head())
print("\n")
print("Amostra - Disney+")
display(df_disney.head())
print("\n")
print("Amostra - Amazon Prime")
display(df_amazon.head())
print("\n")
print("Amostra - Hulu")
display(df_hulu.head())

Amostra - Netflix


show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...



Amostra - Disney+


show_id     type                                             title  \
0      s1    Movie  Duck the Halls: A Mickey Mouse Christmas Special   
1      s2    Movie                            Ernest Saves Christmas   
2      s3    Movie                      Ice Age: A Mammoth Christmas   
3      s4    Movie                        The Queen Family Singalong   
4      s5  TV Show                             The Beatles: Get Back   

                            director  \
0  Alonso Ramirez Ramos, Dave Wasson   
1                        John Cherry   
2                       Karen Disher   
3                    Hamish Hamilton   
4                                NaN   

                                                cast        country  \
0  Chris Diamantopoulos, Tony Anselmo, Tress MacN...            NaN   
1           Jim Varney, Noelle Parker, Douglas Seale            NaN   
2  Raymond Albert Romano, John Leguizamo, Denis L...  United States   
3  Darren Criss, Adam Lambert, Derek Hough, Alexa...            NaN   
4  John Lennon, Paul McCartney, George Harrison, ...            NaN   

          date_added  release_year rating  duration  \
0  November 26, 2021          2016   TV-G    23 min   
1  November 26, 2021          1988     PG    91 min   
2  November 26, 2021          2011   TV-G    23 min   
3  November 26, 2021          2021  TV-PG    41 min   
4  November 25, 2021          2021    NaN  1 Season   

                       listed_in  \
0              Animation, Family   
1                         Comedy   
2      Animation, Comedy, Family   
3                        Musical   
4  Docuseries, Historical, Music   

                                         description  
0   Join Mickey and the gang as they duck the halls!  
1  Santa Claus passes his magic bag to a new St. ...  
2          Sid the Sloth is on Santa's naughty list.  
3               This is real life, not just fantasy!  
4  A three-part documentary from Peter Jackson ca...



Amostra - Amazon Prime


show_id   type                 title        director  \
0      s1  Movie   The Grand Seduction    Don McKellar   
1      s2  Movie  Take Care Good Night    Girish Joshi   
2      s3  Movie  Secrets of Deception     Josh Webber   
3      s4  Movie    Pink: Staying True  Sonia Anderson   
4      s5  Movie         Monster Maker    Giles Foster   

                                                cast         country  \
0     Brendan Gleeson, Taylor Kitsch, Gordon Pinsent          Canada   
1   Mahesh Manjrekar, Abhay Mahajan, Sachin Khedekar           India   
2  Tom Sizemore, Lorenzo Lamas, Robert LaSardo, R...   United States   
3  Interviews with: Pink, Adele, Beyoncé, Britney...   United States   
4  Harry Dean Stanton, Kieran O'Brien, George Cos...  United Kingdom   

       date_added  release_year rating duration                listed_in  \
0  March 30, 2021          2014    NaN  113 min            Comedy, Drama   
1  March 30, 2021          2018    13+  110 min     Drama, International   
2  March 30, 2021          2017    NaN   74 min  Action, Drama, Suspense   
3  March 30, 2021          2014    NaN   69 min              Documentary   
4  March 30, 2021          1989    NaN   45 min           Drama, Fantasy   

                                         description  
0  A small fishing village must procure a local d...  
1  A Metro Family decides to fight a Cyber Crimin...  
2  After a man discovers his wife is cheating on ...  
3  Pink breaks the mold once again, bringing her ...  
4  Teenage Matt Banting wants to work with a famo...



Amostra - Hulu


show_id   type                           title director  cast country  \
0      s1  Movie  Ricky Velez: Here's Everything      NaN   NaN     NaN   
1      s2  Movie                    Silent Night      NaN   NaN     NaN   
2      s3  Movie                    The Marksman      NaN   NaN     NaN   
3      s4  Movie                            Gaia      NaN   NaN     NaN   
4      s5  Movie                        Settlers      NaN   NaN     NaN   

         date_added  release_year rating duration                  listed_in  \
0  October 24, 2021          2021  TV-MA      NaN           Comedy, Stand Up   
1  October 23, 2021          2020    NaN   94 min     Crime, Drama, Thriller   
2  October 23, 2021          2021  PG-13  108 min           Action, Thriller   
3  October 22, 2021          2021      R   97 min                     Horror   
4  October 22, 2021          2021    NaN  104 min  Science Fiction, Thriller   

                                         description  
0  ​Comedian Ricky Velez bares it all with his ho...  
1  Mark, a low end South London hitman recently r...  
2  A hardened Arizona rancher tries to protect an...  
3  A forest ranger and two survivalists with a cu...  
4  Mankind's earliest settlers on the Martian fro...

In [5]:
# Informações gerais dos datasets
print("Informações - Netflix")
df_netflix.info()
print("\n")
print("Informações - Disney+")
df_disney.info()
print("\n")
print("Informações - Amazon Prime")
df_amazon.info()
print("\n")
print("Informações - Hulu")
df_hulu.info()

Informações - Netflix
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


Informações - Disney+
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1450 entries, 0 to 1449
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id   

# Identificação e correção de problemas gerais

## Tratamento de dados nulos

In [6]:
# Verificando valores nulos
print("Valores nulos - Netflix")
print(df_netflix.isnull().sum())
print("\n")
print("Valores nulos - Disney+")
print(df_disney.isnull().sum())
print("\n")
print("Valores nulos - Amazon Prime")
print(df_amazon.isnull().sum())
print("\n")
print("Valores nulos - Hulu")
print(df_hulu.isnull().sum())
print("\n")

Valores nulos - Netflix
show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64


Valores nulos - Disney+
show_id           0
type              0
title             0
director        473
cast            190
country         219
date_added        3
release_year      0
rating            3
duration          0
listed_in         0
description       0
dtype: int64


Valores nulos - Amazon Prime
show_id            0
type               0
title              0
director        2083
cast            1233
country         8996
date_added      9513
release_year       0
rating           337
duration           0
listed_in          0
description        0
dtype: int64


Valores nulos - Hulu
show_id            0
type               0
title              0
director        3070
cast            3073
country 

In [7]:
# Tratamento de dados nulos - Netflix
df_netflix[["director", "cast", "country"]] = df_netflix[["director", "cast", "country"]].fillna("anonymous")
df_netflix['duration'] = df_netflix['duration'].fillna('unknown')
df_netflix['rating'] = df_netflix['rating'].fillna('UNRATED')

In [8]:
# Tratamento de dados nulos - Disney+
df_disney[["director", "cast", "country"]] = df_disney[["director", "cast", "country"]].fillna("anonymous")
df_disney['rating'] = df_disney['rating'].fillna('NR')

In [9]:
# Tratamento de dados nulos - Amazon Prime
df_amazon[["director", "cast", "country"]] = df_amazon[["director", "cast", "country"]].fillna("anonymous")
df_amazon['rating'] = df_amazon['rating'].fillna('UNRATED')

In [10]:
# Tratamento de dados nulos - Hulu
df_hulu[["director", "cast", "country"]] = df_hulu[["director", "cast", "country"]].fillna("anonymous")
df_hulu['duration'] = df_hulu['duration'].fillna('unknown')
df_hulu['rating'] = df_hulu['rating'].fillna('UNRATED')
df_hulu['description'] = df_hulu['description'].fillna('unknown')

**Tratamento dos valores nulos da coluna "data_added"**

Para essa coluna pode-se adotar diferentes abordagens

*Preencimento com uma data aproximada*
- Se o título tem um ano de lançamento (release_year) conhecido, pode-se assumir que ele foi adicionado no mesmo ano. Entretando esse abordagem assume que todos os títulos foram adicionados ao catálogo no mesmo ano de lançamento, o que não é sempre verdadeiro.

*Criar uma categoria "Desconhecido" (unknown)*
- Preencher essa coluna como desconhecido mantém os dados consistentes porque evita a introdução de vieses ao adicionar uma data aproximada, todavia não permite análises baseadas na data.

*Exlcuir registros com dados nulos*
- Caso os valores nulos na coluna fossem pequenos essa seria uma boa opção, mas ela não é ideal para o dataset da Amazon Prime, haja vista que tem 9513 sem registros.


**O que será feito**

*Preenchimento com "unknown"*
- Preencher todos os valores com "unknown" é a abordagem mais consistente nesse caso, porque evita o viés e mantém a integridade dos dados para análises futuras.

In [11]:
# Netflix
df_netflix["date_added"] = df_netflix["date_added"].fillna("unknown")

# Disney+
df_disney["date_added"] = df_disney["date_added"].fillna("unknown")

# Amazon Prime
df_amazon["date_added"] = df_amazon["date_added"].fillna("unknown")

# Hulu
df_hulu["date_added"] = df_hulu["date_added"].fillna("unknown")

In [12]:
print("Valores nulos - Netflix")
print(df_netflix.isnull().sum())
print("\n")
print("Valores nulos - Disney+")
print(df_disney.isnull().sum())
print("\n")
print("Valores nulos - Amazon Prime")
print(df_amazon.isnull().sum())
print("\n")
print("Valores nulos - Hulu")
print(df_hulu.isnull().sum())
print("\n")

Valores nulos - Netflix
show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description     0
dtype: int64


Valores nulos - Disney+
show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description     0
dtype: int64


Valores nulos - Amazon Prime
show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description     0
dtype: int64


Valores nulos - Hulu
show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description

## Remoção de duplicatas

In [13]:
# Verificando se há duplicatas nos datasets

# Netflix
duplicated_netflix = df_netflix.duplicated()
print(f"Netflix - Total de duplicatas: {duplicated_netflix.sum()}")

# Disney+
duplicated_disney = df_disney.duplicated()
print(f"Disney - Total de duplicatas: {duplicated_disney.sum()}")

# Amazon Prime
duplicated_amazon = df_amazon.duplicated()
print(f"Amazon - Total de duplicatas: {duplicated_amazon.sum()}")

# Hulu
duplicated_hulu = df_hulu.duplicated()
print(f"Hulu - Total de duplicatas: {duplicated_hulu.sum()}")

Netflix - Total de duplicatas: 0
Disney - Total de duplicatas: 0
Amazon - Total de duplicatas: 0
Hulu - Total de duplicatas: 0


## Conversão de tipos de dados

In [14]:
# Convertendo a coluna (relaease_year) para datetime

# Netflix
df_netflix["release_year"] = pd.to_datetime(df_netflix["release_year"], format="%Y", errors="coerce").dt.year

# Disney+
df_disney["release_year"] = pd.to_datetime(df_disney["release_year"], format="%Y", errors="coerce").dt.year

# Amazon Prime
df_amazon["release_year"] = pd.to_datetime(df_amazon["release_year"], format="%Y", errors="coerce").dt.year

# Hulu
df_hulu["release_year"] = pd.to_datetime(df_hulu["release_year"], format="%Y", errors="coerce").dt.year

In [15]:
# Informações gerais dos datasets
print("Informações - Netflix")
df_netflix.info()
print("\n")
print("Informações - Disney+")
df_disney.info()
print("\n")
print("Informações - Amazon Prime")
df_amazon.info()
print("\n")
print("Informações - Hulu")
df_hulu.info()

Informações - Netflix
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      8807 non-null   object
 4   cast          8807 non-null   object
 5   country       8807 non-null   object
 6   date_added    8807 non-null   object
 7   release_year  8807 non-null   int32 
 8   rating        8807 non-null   object
 9   duration      8807 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int32(1), object(11)
memory usage: 791.4+ KB


Informações - Disney+
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1450 entries, 0 to 1449
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id   

In [16]:
# Exibir uma amostra dos dados
print("Amostra - Netflix")
display(df_netflix.head())
print("\n")
print("Amostra - Disney+")
display(df_disney.head())
print("\n")
print("Amostra - Amazon Prime")
display(df_amazon.head())
print("\n")
print("Amostra - Hulu")
display(df_hulu.head())

Amostra - Netflix


show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water        anonymous   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans        anonymous   
4      s5  TV Show           Kota Factory        anonymous   

                                                cast        country  \
0                                          anonymous  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...      anonymous   
3                                          anonymous      anonymous   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...



Amostra - Disney+


show_id     type                                             title  \
0      s1    Movie  Duck the Halls: A Mickey Mouse Christmas Special   
1      s2    Movie                            Ernest Saves Christmas   
2      s3    Movie                      Ice Age: A Mammoth Christmas   
3      s4    Movie                        The Queen Family Singalong   
4      s5  TV Show                             The Beatles: Get Back   

                            director  \
0  Alonso Ramirez Ramos, Dave Wasson   
1                        John Cherry   
2                       Karen Disher   
3                    Hamish Hamilton   
4                          anonymous   

                                                cast        country  \
0  Chris Diamantopoulos, Tony Anselmo, Tress MacN...      anonymous   
1           Jim Varney, Noelle Parker, Douglas Seale      anonymous   
2  Raymond Albert Romano, John Leguizamo, Denis L...  United States   
3  Darren Criss, Adam Lambert, Derek Hough, Alexa...      anonymous   
4  John Lennon, Paul McCartney, George Harrison, ...      anonymous   

          date_added  release_year rating  duration  \
0  November 26, 2021          2016   TV-G    23 min   
1  November 26, 2021          1988     PG    91 min   
2  November 26, 2021          2011   TV-G    23 min   
3  November 26, 2021          2021  TV-PG    41 min   
4  November 25, 2021          2021     NR  1 Season   

                       listed_in  \
0              Animation, Family   
1                         Comedy   
2      Animation, Comedy, Family   
3                        Musical   
4  Docuseries, Historical, Music   

                                         description  
0   Join Mickey and the gang as they duck the halls!  
1  Santa Claus passes his magic bag to a new St. ...  
2          Sid the Sloth is on Santa's naughty list.  
3               This is real life, not just fantasy!  
4  A three-part documentary from Peter Jackson ca...



Amostra - Amazon Prime


show_id   type                 title        director  \
0      s1  Movie   The Grand Seduction    Don McKellar   
1      s2  Movie  Take Care Good Night    Girish Joshi   
2      s3  Movie  Secrets of Deception     Josh Webber   
3      s4  Movie    Pink: Staying True  Sonia Anderson   
4      s5  Movie         Monster Maker    Giles Foster   

                                                cast         country  \
0     Brendan Gleeson, Taylor Kitsch, Gordon Pinsent          Canada   
1   Mahesh Manjrekar, Abhay Mahajan, Sachin Khedekar           India   
2  Tom Sizemore, Lorenzo Lamas, Robert LaSardo, R...   United States   
3  Interviews with: Pink, Adele, Beyoncé, Britney...   United States   
4  Harry Dean Stanton, Kieran O'Brien, George Cos...  United Kingdom   

       date_added  release_year   rating duration                listed_in  \
0  March 30, 2021          2014  UNRATED  113 min            Comedy, Drama   
1  March 30, 2021          2018      13+  110 min     Drama, International   
2  March 30, 2021          2017  UNRATED   74 min  Action, Drama, Suspense   
3  March 30, 2021          2014  UNRATED   69 min              Documentary   
4  March 30, 2021          1989  UNRATED   45 min           Drama, Fantasy   

                                         description  
0  A small fishing village must procure a local d...  
1  A Metro Family decides to fight a Cyber Crimin...  
2  After a man discovers his wife is cheating on ...  
3  Pink breaks the mold once again, bringing her ...  
4  Teenage Matt Banting wants to work with a famo...



Amostra - Hulu


show_id   type                           title   director       cast  \
0      s1  Movie  Ricky Velez: Here's Everything  anonymous  anonymous   
1      s2  Movie                    Silent Night  anonymous  anonymous   
2      s3  Movie                    The Marksman  anonymous  anonymous   
3      s4  Movie                            Gaia  anonymous  anonymous   
4      s5  Movie                        Settlers  anonymous  anonymous   

     country        date_added  release_year   rating duration  \
0  anonymous  October 24, 2021          2021    TV-MA  unknown   
1  anonymous  October 23, 2021          2020  UNRATED   94 min   
2  anonymous  October 23, 2021          2021    PG-13  108 min   
3  anonymous  October 22, 2021          2021        R   97 min   
4  anonymous  October 22, 2021          2021  UNRATED  104 min   

                   listed_in  \
0           Comedy, Stand Up   
1     Crime, Drama, Thriller   
2           Action, Thriller   
3                     Horror   
4  Science Fiction, Thriller   

                                         description  
0  ​Comedian Ricky Velez bares it all with his ho...  
1  Mark, a low end South London hitman recently r...  
2  A hardened Arizona rancher tries to protect an...  
3  A forest ranger and two survivalists with a cu...  
4  Mankind's earliest settlers on the Martian fro...

# Identificação e correção de problemas particulares

## Netflix

In [17]:
# Visualizando as informações que estão contidas nas colunas
for column in df_netflix.columns:
    print(f"Coluna: {column}")
    print(df_netflix[column].unique())
    print("-"*50)

Coluna: show_id
['s1' 's2' 's3' ... 's8805' 's8806' 's8807']
--------------------------------------------------
Coluna: type
['Movie' 'TV Show']
--------------------------------------------------
Coluna: title
['Dick Johnson Is Dead' 'Blood & Water' 'Ganglands' ... 'Zombieland'
 'Zoom' 'Zubaan']
--------------------------------------------------
Coluna: director
['Kirsten Johnson' 'anonymous' 'Julien Leclercq' ... 'Majid Al Ansari'
 'Peter Hewitt' 'Mozez Singh']
--------------------------------------------------
Coluna: cast
['anonymous'
 'Ama Qamata, Khosi Ngema, Gail Mabalane, Thabang Molaba, Dillon Windvogel, Natasha Thahane, Arno Greeff, Xolile Tshabalala, Getmore Sithole, Cindy Mahlangu, Ryle De Morny, Greteli Fincham, Sello Maake Ka-Ncube, Odwa Gwanya, Mekaila Mathys, Sandi Schultz, Duane Williams, Shamilla Miller, Patrick Mofokeng'
 'Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabiha Akkari, Sofia Lesaffre, Salim Kechiouche, Noureddine Farihi, Geert Van Rampelberg, Bakary Diombera

In [18]:
# Verificando a padronização dos ids na coluna "show_id"
pattern = r"^s\d+$"
invalid_rows = df_netflix[~df_netflix["show_id"].str.match(pattern, na=False)]
print("Registros que não seguem o padrão 'sXXX'")
invalid_rows

Registros que não seguem o padrão 'sXXX'


Empty DataFrame
Columns: [show_id, type, title, director, cast, country, date_added, release_year, rating, duration, listed_in, description]
Index: []

In [19]:
# Verificando se há valores fora do esperado na coluna "type"
expected_values = ["Movie", "TV Show"]
invalid_type_rows = df_netflix[~df_netflix["type"].isin(expected_values)]
print("Registros com valores inesperados na coluna 'type':")
invalid_type_rows

Registros com valores inesperados na coluna 'type':


Empty DataFrame
Columns: [show_id, type, title, director, cast, country, date_added, release_year, rating, duration, listed_in, description]
Index: []

In [20]:
# Verificando as linhas que estão com valores incorretos na coluna "rating"
duration_pattern = r"^\d+\smin$"
invalid_rating_rows = df_netflix[df_netflix["rating"].str.match(duration_pattern, na=False)]
print("Linhas com valores inconsistentes na coluna 'rating':")
invalid_rating_rows

Linhas com valores inconsistentes na coluna 'rating':


show_id   type                                 title    director  \
5541   s5542  Movie                       Louis C.K. 2017  Louis C.K.   
5794   s5795  Movie                 Louis C.K.: Hilarious  Louis C.K.   
5813   s5814  Movie  Louis C.K.: Live at the Comedy Store  Louis C.K.   

            cast        country          date_added  release_year  rating  \
5541  Louis C.K.  United States       April 4, 2017          2017  74 min   
5794  Louis C.K.  United States  September 16, 2016          2010  84 min   
5813  Louis C.K.  United States     August 15, 2016          2015  66 min   

     duration listed_in                                        description  
5541  unknown    Movies  Louis C.K. muses on religion, eternal love, gi...  
5794  unknown    Movies  Emmy-winning comedy writer Louis C.K. brings h...  
5813  unknown    Movies  The comic puts his trademark hilarious/thought...

In [21]:
# Movendo os valores de "rating" para a coluna "duration"
invalid_rows = df_netflix["rating"].str.match(duration_pattern, na=False)
df_netflix.loc[invalid_rows, "duration"] = df_netflix.loc[invalid_rows, "rating"]
df_netflix.loc[invalid_rows, "rating"] = 'UNRATED' 
print("Verificando a correção:")
print(df_netflix["rating"].unique())

Verificando a correção:
['PG-13' 'TV-MA' 'PG' 'TV-14' 'TV-PG' 'TV-Y' 'TV-Y7' 'R' 'TV-G' 'G'
 'NC-17' 'UNRATED' 'NR' 'TV-Y7-FV' 'UR']


## Disney +

In [22]:
# Visualizando as informações que estão contidas nas colunas
for column in df_disney.columns:
    print(f"Coluna: {column}")
    print(df_disney[column].unique())
    print("-"*50)

Coluna: show_id
['s1' 's2' 's3' ... 's1448' 's1449' 's1450']
--------------------------------------------------
Coluna: type
['Movie' 'TV Show']
--------------------------------------------------
Coluna: title
['Duck the Halls: A Mickey Mouse Christmas Special'
 'Ernest Saves Christmas' 'Ice Age: A Mammoth Christmas' ...
 'Eddie the Eagle' 'Bend It Like Beckham'
 'Captain Sparky vs. The Flying Saucers']
--------------------------------------------------
Coluna: director
['Alonso Ramirez Ramos, Dave Wasson' 'John Cherry' 'Karen Disher'
 'Hamish Hamilton' 'anonymous' 'Liz Garbus' 'Kirk R. Thatcher'
 'John Gleim' 'Leslie Iwerks' 'McKenna Harris' 'Kevin Lima'
 'Patrick Osborne' 'Chris Buck, Jennifer Lee' 'Lauren MacMullan'
 'Dan Mazer' 'Jaume Collet-Serra' 'John Kahrs' 'Destin Daniel Cretton'
 'Manjari Makijany' 'Nathan Greno, Byron Howard'
 'Stevie Wermers-Skelton, Kevin Deters' 'Roger Allers' 'David Clair'
 'David Silverman' 'Dean Wellins' 'Mike Mitchell' 'Brian Levant'
 'Alex Zamm' 'Kev

## Amazon Prime

In [23]:
# Visualizando as informações que estão contidas nas colunas
for column in df_amazon.columns:
    print(f"Coluna: {column}")
    print(df_amazon[column].unique())
    print("-"*50)

Coluna: show_id
['s1' 's2' 's3' ... 's9666' 's9667' 's9668']
--------------------------------------------------
Coluna: type
['Movie' 'TV Show']
--------------------------------------------------
Coluna: title
['The Grand Seduction' 'Take Care Good Night' 'Secrets of Deception' ...
 'Outpost' 'Maradona: Blessed Dream' 'Harry Brown']
--------------------------------------------------
Coluna: director
['Don McKellar' 'Girish Joshi' 'Josh Webber' ...
 'John-Paul Davidson, Stephen Warbeck' 'Emily Skye' 'Steve Barker']
--------------------------------------------------
Coluna: cast
['Brendan Gleeson, Taylor Kitsch, Gordon Pinsent'
 'Mahesh Manjrekar, Abhay Mahajan, Sachin Khedekar'
 'Tom Sizemore, Lorenzo Lamas, Robert LaSardo, Richard Jones, Yancey Arias, Noel Gugliemi'
 ...
 'Ray Stevenson, Julian Wadham, Richard Brake, Michael Smiley, Enoch Frost'
 'Esteban Recagno, Ezequiel Stremiz, Luciano Vittori, Pablo Rodríguez Albi, Leonard Kunz, Juanma Muniagurria, Julieta Cardinali, Nazareno Case

**Coluna "Rating"**

- Nessa coluna existem dados com 'UNRATED' e 'NOT_RATED' e ambos siginificam que não há classificação. Para seguir o padrão adotado e evitar redundância, aqueles que estão com 'NOT_RATED' serão trocados para 'UNRATED'.

In [24]:
# Substituindo "NOT_RATED" por "UNRATED" na coluna "rating"
df_amazon["rating"] = df_amazon["rating"].replace("NOT_RATE", "UNRATED")
print("Verificando a correção:")
print(df_amazon["rating"].unique())

Verificando a correção:
['UNRATED' '13+' 'ALL' '18+' 'R' 'TV-Y' 'TV-Y7' 'NR' '16+' 'TV-PG' '7+'
 'TV-14' 'TV-NR' 'TV-G' 'PG-13' 'TV-MA' 'G' 'PG' 'NC-17' '16' 'AGES_16_'
 'AGES_18_' 'ALL_AGES']


## Hulu

In [25]:
for column in df_hulu.columns:
    print(f"Coluna: {column}")
    print(df_hulu[column].unique())
    print("-"*50)

Coluna: show_id
['s1' 's2' 's3' ... 's3071' 's3072' 's3073']
--------------------------------------------------
Coluna: type
['Movie' 'TV Show']
--------------------------------------------------
Coluna: title
["Ricky Velez: Here's Everything" 'Silent Night' 'The Marksman' ...
 'The Fades' 'The Twilight Zone' 'Tokyo Magnitude 8.0']
--------------------------------------------------
Coluna: director
['anonymous'
 'Director Jennifer Kent ("The Babadook") tells the sweeping tale of a young woman\'s quest for revenge set in 1825 against the perilous and unforgiving terrain of Tasmania.'
 'Director Gigi Saul Guerrero talks through Culture Shock’s themes – immigration, filmmaking, and latinidad – in this one-on-one chat.'
 'Director Alex Winter goes inside the biggest global corruption scandal in history that was uncovered by hundreds of journalists, working in secret and at enormous risk.']
--------------------------------------------------
Coluna: cast
['anonymous']
-----------------------

In [26]:
# Substituindo "NOT_RATED" por "UNRATED" na coluna "rating"
df_hulu["rating"] = df_hulu["rating"].replace("NOT RATED", "UNRATED")
df_hulu["rating"] = df_hulu["rating"].replace("NR", "UNRATED")
print("Verificando a correção:")
print(df_hulu["rating"].unique())

Verificando a correção:
['TV-MA' 'UNRATED' 'PG-13' 'R' 'TV-14' 'PG' 'TV-PG' 'G' 'TV-G' '2 Seasons'
 'TV-Y' '93 min' '4 Seasons' 'TV-Y7' '136 min' '91 min' '85 min' '98 min'
 '89 min' '94 min' '86 min' '3 Seasons' '121 min' '88 min' '101 min'
 '1 Season' '83 min' '100 min' '95 min' '92 min' '96 min' '109 min'
 '99 min' '75 min' '87 min' '67 min' '104 min' '107 min' '84 min'
 '103 min' '105 min' '119 min' '114 min' '82 min' '90 min' '130 min'
 '110 min' '80 min' '6 Seasons' '97 min' '111 min' '81 min' '49 min'
 '45 min' '41 min' '73 min' '40 min' '36 min' '39 min' '34 min' '47 min'
 '65 min' '37 min' '78 min' '102 min' '129 min' '115 min' '112 min'
 '61 min' '106 min' '76 min' '77 min' '79 min' '157 min' '28 min' '64 min'
 '7 min' '5 min' '6 min' '127 min' '142 min' '108 min' '57 min' '118 min'
 '116 min' '12 Seasons' '71 min']


In [27]:
# Verificando as linhas que estão com valores incorretos sobre tempo de duração para filmes ("XXX min") na coluna "rating"
duration_pattern = r"^\d+\smin$"
invalid_rating_rows = df_hulu[df_hulu["rating"].str.match(duration_pattern, na=False)]
print("Linhas com valores inconsistentes na coluna 'rating':")
invalid_rating_rows

Linhas com valores inconsistentes na coluna 'rating':


show_id   type                                              title  \
108     s109  Movie                                               Maze   
397     s398  Movie                        Gunfight at the O.K. Corral   
495     s496  Movie                    78/52: Hitchcock's Shower Scene   
510     s511  Movie                  Citizen Jane: Battle for the City   
513     s514  Movie                                              Dealt   
...      ...    ...                                                ...   
2785   s2786  Movie  The Beatles: Eight Days a Week - The Touring Y...   
2879   s2880  Movie                                         The Colony   
2907   s2908  Movie                 Ron White: A Little Unprofessional   
2958   s2959  Movie                                 (Sub) Ninja Scroll   
2993   s2994  Movie                     Amy Winehouse: A Final Goodbye   

       director       cast                                   country  \
108   anonymous  anonymous  Ireland, United Kingdom, Germany, Sweden   
397   anonymous  anonymous                             United States   
495   anonymous  anonymous                             United States   
510   anonymous  anonymous                             United States   
513   anonymous  anonymous                             United States   
...         ...        ...                                       ...   
2785  anonymous  anonymous             United States, United Kingdom   
2879  anonymous  anonymous                     Canada, United States   
2907  anonymous  anonymous                             United States   
2958  anonymous  anonymous                                     Japan   
2993  anonymous  anonymous                            United Kingdom   

              date_added  release_year   rating duration  \
108      October 1, 2021          2017   93 min  unknown   
397       August 1, 2021          1957  136 min  unknown   
495         July 1, 2021          2017   91 min  unknown   
510         July 1, 2021          2017   93 min  unknown   
513         July 1, 2021          2017   85 min  unknown   
...                  ...           ...      ...      ...   
2785  September 17, 2016          2016  106 min  unknown   
2879       June 12, 2015          2013   94 min  unknown   
2907     January 1, 2015          2012   80 min  unknown   
2958   December 15, 2013          1993   93 min  unknown   
2993   December 13, 2012          2011   71 min  unknown   

                               listed_in  \
108             Action, Adventure, Crime   
397       Classics, Documentaries, Drama   
495              Documentaries, Thriller   
510                        Documentaries   
513                        Documentaries   
...                                  ...   
2785                Documentaries, Music   
2879           Action, Adventure, Horror   
2907                    Comedy, Stand Up   
2958  Action, Adult Animation, Adventure   
2993                Documentaries, Music   

                                            description  
108   Based on the incredible true story of the 1983...  
397   Dramatization of the legendary battle between ...  
495   Just when you thought it was safe to go back i...  
510   Chronicle of activist Jane Jacobs' battle with...  
513   Prepare to be inspired and amazed by Richard T...  
...                                                 ...  
2785  Academy Award-winner Ron Howard’s highly antic...  
2879  Years ago, winter came and left. Survivors hav...  
2907  Ron White delivers more than an hour of all ne...  
2958  When Jubei saves a young ninja woman from the ...  
2993  A Final Goodbye is an intimate look into the l...  

[230 rows x 12 columns]

In [28]:
# Movendo os valores inconsistentes encontrados na célula anterior para a coluna "duration"
invalid_rows = df_hulu["rating"].str.match(duration_pattern, na=False)
df_hulu.loc[invalid_rows, "duration"] = df_hulu.loc[invalid_rows, "rating"]
df_hulu.loc[invalid_rows, "rating"] = 'UNRATED' 
print("Verificando a correção:")
print(df_hulu["rating"].unique())

Verificando a correção:
['TV-MA' 'UNRATED' 'PG-13' 'R' 'TV-14' 'PG' 'TV-PG' 'G' 'TV-G' '2 Seasons'
 'TV-Y' '4 Seasons' 'TV-Y7' '3 Seasons' '1 Season' '6 Seasons'
 '12 Seasons']


In [29]:
# Verificando as linhas que estão com valores incorretos sobre quantidade temporadas das séries ("XXX Seasons") na coluna "rating"
season_pattern = r"^\d+\sSeason(s)?$"
invalid_rating_rows = df_hulu[df_hulu["rating"].str.match(season_pattern, na=False)]
invalid_rating_rows

show_id   type                           title   director       cast  \
50       s51  Movie  Yashahime: Princess Half-Demon  anonymous  anonymous   
263     s264  Movie             Trolls: TrollsTopia  anonymous  anonymous   
471     s472  Movie                     This Way Up  anonymous  anonymous   
801     s802  Movie      Everything's Gonna Be Okay  anonymous  anonymous   
847     s848  Movie                            Juda  anonymous  anonymous   
...      ...    ...                             ...        ...        ...   
2908   s2909  Movie                  The Wrong Mans  anonymous  anonymous   
2926   s2927  Movie         The Hotwives of Orlando  anonymous  anonymous   
2951   s2952  Movie               All Saints (1998)  anonymous  anonymous   
2955   s2956  Movie                      Getting On  anonymous  anonymous   
2959   s2960  Movie           Packed to the Rafters  anonymous  anonymous   

             country         date_added  release_year      rating duration  \
50             Japan    October 9, 2021          2020   2 Seasons  unknown   
263    United States  September 2, 2021          2020   4 Seasons  unknown   
471   United Kingdom       July 9, 2021          2019   2 Seasons  unknown   
801    United States      April 9, 2021          2020   2 Seasons  unknown   
847           Israel     March 30, 2021          2017   2 Seasons  unknown   
...              ...                ...           ...         ...      ...   
2908  United Kingdom  December 24, 2014          2013   2 Seasons  unknown   
2926   United States      July 15, 2014          2014    1 Season  unknown   
2951       Australia  February 19, 2014          1998  12 Seasons  unknown   
2955  United Kingdom    January 7, 2014          2009   3 Seasons  unknown   
2959       Australia   December 3, 2013          2008   6 Seasons  unknown   

                          listed_in  \
50                            Anime   
263                    Family, Kids   
471    Comedy, Drama, International   
801                   Comedy, Drama   
847        Action, Adventure, Crime   
...                             ...   
2908      Action, Adventure, Comedy   
2926                 Comedy, Sitcom   
2951           Drama, International   
2955  Comedy, International, Sitcom   
2959   Comedy, Drama, International   

                                            description  
50    Long-lost twins Towa and Setsuna reunite after...  
263   Trolls: TrollsTopia! is the next chapter in th...  
471   This Way Up is a comedy drama about moving on,...  
801   Australian twenty-something Nicholas is left t...  
847   Juda is a low-life gambler hustling a living i...  
...                                                 ...  
2908  Lowly office workers Sam (Matthew Baynton) and...  
2926  The reality genre parody examines the glamorou...  
2951  All Saints is an Australian medical drama focu...  
2955  Care for the elderly is the least glamorous ar...  
2959  Packed to the Rafters is an Australian family-...  

[62 rows x 12 columns]

In [30]:
# Movendo os valores inconsistentes encontrados na célula anterior para a coluna "duration"
invalid_rows = df_hulu["rating"].str.match(season_pattern, na=False)
df_hulu.loc[invalid_rows, "duration"] = df_hulu.loc[invalid_rows, "rating"]
df_hulu.loc[invalid_rows, "rating"] = "UNRATED"
print("Verificando a correção:")
print(df_hulu["rating"].unique())

Verificando a correção:
['TV-MA' 'UNRATED' 'PG-13' 'R' 'TV-14' 'PG' 'TV-PG' 'G' 'TV-G' 'TV-Y'
 'TV-Y7']


# Integração dos datasets

## Criando colunas para identificar a plataforma

In [31]:
df_netflix["platform"] = "Netflix"
df_disney["platform"] = "Disney+"
df_amazon["platform"] = "Amazon Prime"
df_hulu["platform"] = "Hulu"

## Combinando os datasets

In [32]:
df_combined = pd.concat([df_netflix, df_disney, df_amazon, df_hulu])

## Salvando o dataset limpo

In [33]:
df_combined.to_excel("streaming_platforms.xlsx", index = False)

In [34]:
#df_netflix.to_excel("df_netflix.xlsx", index = False)
#df_disney.to_excel("df_disney.xlsx", index = False)
#df_amazon.to_excel("df_amazon.xlsx", index = False)
#df_hulu.to_excel("df_hulu.xlsx", index = False)

In [35]:
print(df_combined["rating"].unique())

['PG-13' 'TV-MA' 'PG' 'TV-14' 'TV-PG' 'TV-Y' 'TV-Y7' 'R' 'TV-G' 'G'
 'NC-17' 'UNRATED' 'NR' 'TV-Y7-FV' 'UR' '13+' 'ALL' '18+' '16+' '7+'
 'TV-NR' '16' 'AGES_16_' 'AGES_18_' 'ALL_AGES']
